lib/spack/spack/database.py

```diff
@@ -206,9 +223,12 @@ def from_dict(cls, spec, dictionary):
         d.pop('spec', None)
 
         # Old databases may have "None" for path for externals
-        if d['path'] == 'None':
+        if 'path' not in d or d['path'] == 'None':
             d['path'] = None
 
+        if 'installed' not in d:
+            d['installed'] = False
```

Comment:
This bit concerns me. Can you explain what's going on here a bit?

Reply:
When we use the database to write out the buildcache index, one of the fields we leave out is the 'installed' field.  So when we read back in, we have to be ok if the `installed` field is missing at that point too.  To me it seems that if the field is missing, it's safe to say the spec is not installed.  But let me know if you disagree.

Reply:
FYI, without lines 229-230, I get the following stack trace when I try to read the buildcache into the database:

```
Traceback (most recent call last):
  File "/Users/scott/projects/spack/lib/spack/spack/database.py", line 777, in _read_from_file
    data[hash_key] = InstallRecord.from_dict(spec, rec)
  File "/Users/scott/projects/spack/lib/spack/spack/database.py", line 232, in from_dict
    return InstallRecord(spec, **d)
TypeError: __init__() missing 1 required positional argument: 'installed'
```